linear_operator/operators/permutation_linear_operator.py

```diff
@@ -0,0 +1,168 @@
+            sorted_perm, inv_perm = perm.sort(dim=-1)
+
+        if validate_args:
+            if torch.is_floating_point(sorted_perm) or torch.is_complex(sorted_perm):
+                raise ValueError("perm does not have integer elements.")
+
+            for i, si in enumerate(sorted_perm):
+                if (sorted_perm[..., i] != i).any():
+                    raise ValueError(
+                        f"Invalid perm-inv_perm input, index {i} missing or not at "
+                        f"correct index for permutation with {perm.shape = }."
+                    )
+
+        self.perm = perm
+        self.inv_perm = inv_perm
+        super().__init__(perm, inv_perm, validate_args=validate_args)
+
+    def _matmul(self, rhs: Tensor) -> Tensor:
+        # input rhs is guaranteed to be at least two-dimensional due to matmul implementation
+        self._matmul_check_shape(rhs)
+
+        # batch broadcasting logic
+        batch_shape = self._matmul_batch_shape(rhs)
+        expanded_rhs = rhs.expand(*batch_shape, *rhs.shape[-2:])
+        ndim = expanded_rhs.ndim
+
+        batch_indices = self._batch_indexing_helper(batch_shape)
+        batch_indices = tuple(index[..., None] for index in batch_indices)  # expanding to non-batch dimensions
+        perm_indices = self.perm[..., None]
+        final_indices = torch.arange(rhs.shape[-1])[(None,) * (ndim - 1) + (slice(None),)]
+        indices = batch_indices + (perm_indices, final_indices)
+        return expanded_rhs[indices]
+
+    def _batch_indexing_helper(self, batch_shape: torch.Size) -> Tuple:
+        indices = []
+        for i, n in enumerate(batch_shape):
+            indices.append(
+                torch.arange(n)[
+                    (None,) * i
+                    + (slice(None),)
+                    + (None,) * (len(batch_shape) - i - 1)
+                    + (None,)  # adding one non-batch dimension
+                ]
+            )
+        return tuple(indices)
+
+    def _size(self) -> torch.Size:
+        return torch.Size((*self.perm.shape, self.perm.shape[-1]))
+
+    def _transpose_nonbatch(self):
+        return PermutationLinearOperator(perm=self.inv_perm, inv_perm=self.perm, validate_args=False)
+
+    def to_sparse(self):
+        # crow_indices[i] is index where values of row i begin
+        return torch.sparse_csr_tensor(
+            crow_indices=torch.arange(self.shape[-1] + 1).expand(*self.batch_shape, -1).contiguous(),
+            col_indices=self.perm,
+            values=torch.ones_like(self.perm),
+        )
+
+
+class TransposePermutationLinearOperator(AbstractPermutationLinearOperator):
+    r"""LinearOperator that represents a permutation matrix `P` with O(1) memory.
+    In particular, P satisfies
+
+        `P @ X.flatten(-2, -1) = X.transpose(-2, -1).flatten(-2, -1)`,
+
+    where `X` is an `m x m` matrix and P has size `n x n` where `n = m^2`.
+
+    Args:
+        - m: dimension on which the transpose operation is taking place. The size of
```

Comment:
If we decide to keep `batch_shape`, I should add it to the doc string here.

Reply:
Hmm the only thing batch shape does is to give it a specific `size`? I don't think that's necessary so long as the operator is properly broadcastable (which it seems it is). Not sure if we want to ever represent the permutation with a specific batch shape. @gpleiss any opinions here?